File: smoke_tests/cpp/ssrf.cpp
// SSRF (Server-Side Request Forgery) vulnerabilities in C++
#include <iostream>
#include <string>
#include <curl/curl.h>

// Test 1: Direct URL from user input
void fetch_url(const std::string& url) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: User-controlled URL
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

// Test 2: URL parameter used in request
void fetch_image(const std::string& image_url) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: image_url from user
        curl_easy_setopt(curl, CURLOPT_URL, image_url.c_str());
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

// Test 3: Partial URL construction
void fetch_from_host(const std::string& hostname) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: hostname from user
        std::string url = "http://" + hostname + "/api/data";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

// Test 4: Port injection
void fetch_with_port(const std::string& host, const std::string& port) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: port from user, can access internal services
        std::string url = "http://internal-server:" + port + "/";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

// Test 5: URL in redirect
void handle_redirect(const std::string& redirect_url) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: Following redirects to user-controlled URL
        curl_easy_setopt(curl, CURLOPT_URL, redirect_url.c_str());
        curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
        curl_easy_setopt(curl, CURLOPT_MAXREDIRS, 10L);
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

// Test 6: DNS rebinding vulnerability
void fetch_external(const std::string& domain) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: DNS can resolve to internal IP after validation
        std::string url = "http://" + domain + "/data";
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

// Test 7: File protocol SSRF
void fetch_resource(const std::string& resource_url) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: Could be file:///etc/passwd
        curl_easy_setopt(curl, CURLOPT_URL, resource_url.c_str());
        curl_easy_setopt(curl, CURLOPT_PROTOCOLS, CURLPROTO_ALL);  // Allows file://
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

// Test 8: Webhook URL
void send_webhook(const std::string& webhook_url, const std::string& data) {
    CURL* curl = curl_easy_init();
    if (curl) {
        // VULNERABLE: webhook_url from user configuration
        curl_easy_setopt(curl, CURLOPT_URL, webhook_url.c_str());
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data.c_str());
        curl_easy_perform(curl);
        curl_easy_cleanup(curl);
    }
}

File: smoke_tests/swift/ssrf.swift
// Server-Side Request Forgery (SSRF) vulnerabilities in Swift
import Foundation

class SsrfVulnerabilities {

    // Test 1: URLSession with user-controlled URL
    func fetchUrl(urlString: String, completion: @escaping (Data?) -> Void) {
        guard let url = URL(string: urlString) else {
            completion(nil)
            return
        }
        // VULNERABLE: User-controlled URL
        URLSession.shared.dataTask(with: url) { data, _, _ in
            completion(data)
        }.resume()
    }

    // Test 2: Partial URL construction
    func fetchFromHost(host: String, completion: @escaping (Data?) -> Void) {
        // VULNERABLE: User controls hostname
        let urlString = "http://\(host)/api/data"
        guard let url = URL(string: urlString) else {
            completion(nil)
            return
        }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            completion(data)
        }.resume()
    }

    // Test 3: Port scanning via SSRF
    func checkPort(port: Int, completion: @escaping (Bool) -> Void) {
        // VULNERABLE: User controls port
        let urlString = "http://internal-server:\(port)/"
        guard let url = URL(string: urlString) else {
            completion(false)
            return
        }

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 5
        let session = URLSession(configuration: config)

        session.dataTask(with: url) { _, response, error in
            completion(error == nil)
        }.resume()
    }

    // Test 4: Image proxy
    func fetchImage(imageUrl: String, completion: @escaping (Data?) -> Void) {
        guard let url = URL(string: imageUrl) else {
            completion(nil)
            return
        }
        // VULNERABLE: Can fetch internal resources
        URLSession.shared.dataTask(with: url) { data, _, _ in
            completion(data)
        }.resume()
    }

    // Test 5: Webhook URL
    func sendWebhook(webhookUrl: String, data: Data, completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: webhookUrl) else {
            completion(false)
            return
        }
        // VULNERABLE: User-controlled webhook destination
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.httpBody = data

        URLSession.shared.dataTask(with: request) { _, _, error in
            completion(error == nil)
        }.resume()
    }

    // Test 6: DNS rebinding
    func fetchExternal(domain: String, completion: @escaping (Data?) -> Void) {
        // VULNERABLE: DNS can resolve to internal IP
        let urlString = "http://\(domain)/data"
        guard let url = URL(string: urlString) else {
            completion(nil)
            return
        }
        URLSession.shared.dataTask(with: url) { data, _, _ in
            completion(data)
        }.resume()
    }

    // Test 7: File protocol SSRF
    func readResource(uri: String, completion: @escaping (Data?) -> Void) {
        guard let url = URL(string: uri) else {
            completion(nil)
            return
        }
        // VULNERABLE: Could be file:///etc/passwd
        URLSession.shared.dataTask(with: url) { data, _, _ in
            completion(data)
        }.resume()
    }

    // Test 8: Redirect following
    func fetchWithRedirect(urlString: String, completion: @escaping (Data?) -> Void) {
        guard let url = URL(string: urlString) else {
            completion(nil)
            return
        }

        let config = URLSessionConfiguration.default
        // VULNERABLE: Following redirects to internal resources
        let session = URLSession(configuration: config)

        session.dataTask(with: url) { data, _, _ in
            completion(data)
        }.resume()
    }

    // Test 9: URL from user input in API call
    func callExternalApi(endpoint: String, completion: @escaping (Data?) -> Void) {
        // VULNERABLE: Endpoint from user
        guard let url = URL(string: endpoint) else {
            completion(nil)
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        URLSession.shared.dataTask(with: request) { data, _, _ in
            completion(data)
        }.resume()
    }

    // Test 10: URL components manipulation
    func fetchWithPath(basePath: String, completion: @escaping (Data?) -> Void) {
        var components = URLComponents(string: "http://api.example.com")!
        // VULNERABLE: Path from user can include /../ or @
        components.path = basePath

        guard let url = components.url else {
            completion(nil)
            return
        }

        URLSession.shared.dataTask(with: url) { data, _, _ in
            completion(data)
        }.resume()
    }
}
